functoria/lib/functoria.ml

```diff
@@ -250,3 +247,161 @@ module type KEY =
+          Codegen.newline fmt;
+          let keys = Key.Set.of_list @@ Info.keys i in
+          let pp_var k = Key.serialize (Info.context i) k in
+          Fmt.pf fmt "@[<v>%a@]@." (Fmt.iter Key.Set.iter pp_var) keys;
+          let runvars = Key.Set.elements (Key.filter_stage `Run keys) in
+          let pp_runvar ppf v = Fmt.pf ppf "%s_t" (Key.ocaml_name v) in
+          let pp_names ppf v = Fmt.pf ppf "%S" (Key.name v) in
+          Codegen.append fmt "let runtime_keys = List.combine %a %a"
+            Fmt.Dump.(list pp_runvar) runvars Fmt.Dump.(list pp_names) runvars;
+          Codegen.newline fmt;
+          R.ok ())
+
+  let clean _i = Bos.OS.Path.delete file
+
+  let name = "key"
+
+end
+
+let keys (argv: argv impl) = impl @@ object
+    inherit base_configurable
+    method ty = job
+    method name = Keys.name
+    method module_name = Key.module_name
+    method !configure = Keys.configure
+    method !clean = Keys.clean
+    method !packages = Key.pure [package "functoria-runtime"]
+    method !deps = [ abstract argv ]
+    method !connect info modname = function
+      | [ argv ] ->
+        Fmt.strf
+          "return (Functoria_runtime.with_argv (List.map fst %s.runtime_keys) %S %s)"
+          modname (Info.name info) argv
+      | _ -> failwith "The keys connect should receive exactly one argument."
+  end
+
+(* Module emiting a file containing all the build information. *)
+
+type info = Info
+let info = Type Info
+
+let pp_libraries fmt l =
+  Fmt.pf fmt "[@ %a]"
+    Fmt.(iter ~sep:(unit ";@ ") List.iter @@ fmt "%S") l
+
+let pp_packages fmt l =
+  Fmt.pf fmt "[@ %a]"
+    Fmt.(iter ~sep:(unit ";@ ") List.iter @@
+         (fun fmt (n, v) -> pf fmt "%S, %S" n v)
+        ) l
+
+let pp_dump_pkgs module_name fmt (name, pkg, libs) =
+  Fmt.pf fmt
+    "%s.{@ name = %S;@ \
+     @[<v 2>packages = %a@]@ ;@ @[<v 2>libraries = %a@]@ }"
+    module_name name
+    pp_packages (String.Map.bindings pkg)
+    pp_libraries (String.Set.elements libs)
+
+let app_info ?(type_modname="Functoria_info")  ?(gen_modname="Info_gen") () =
+  impl @@ object
+    inherit base_configurable
+    method ty = info
+    method name = "info"
+    val file = Fpath.(v (String.Ascii.lowercase gen_modname) + "ml")
+    method module_name = gen_modname
+    method !packages = Key.pure [package "functoria-runtime"]
+    method !connect _ modname _ = Fmt.strf "return %s.info" modname
+
+    method !clean _i = Bos.OS.Path.delete file
+
+    method !configure _i = Ok ()
+
+    method !build i =
+      Log.info (fun m -> m "Generating: %a" Fpath.pp file);
+      (* this used to call 'opam list --rec ..', but that leads to
+         non-reproducibility, since this uses the opam CUDF solver which
+         drops some packages (which are in the repositories configured for the
+         switch), see https://github.com/mirage/functoria/pull/189 for further
+         discussion on this before changing the code below.  *)
```

Comment:
you may want to update the comment why `--resolve` is used now -- I tested the `--resolve` command manually below, and it seems to work fine for me